### PR TITLE
Chapter 8: Basic login

### PR DESCRIPTION
**Pull request by ainulfarizah**
ready

### CODE REVIEW

**Comment by ainulfarizah** [comment]
ready

File: app/assets/stylesheets/sessions.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Sessions controller here.
```
Comment (ghost) [inline app/assets/stylesheets/sessions.scss]: please deleted this file

File: app/views/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<%= provide :title, t("sign_up") %>
```
Comment (ghost) [inline app/views/new.html.erb]:
```suggestion
<%= provide :title,"Sign Up") %>
```
If you don't use i18n

File: test/controllers/sessions_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+require "test_helper"
+
```
Comment (ghost) [inline test/controllers/sessions_controller_test.rb]: please delete this file

File: app/views/sessions/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,14 @@
+<% provide :title, t("log_in") %>
```
Comment (anlvq-1058) [inline app/views/sessions/new.html.erb:1]: fix i18n syntax

**Review by ghost — COMMENTED** [review]
cmt

File: app/views/new.html.erb (hunk 2 above)
Comment (ainulfarizah) [reply app/views/new.html.erb]: i already put sign up in en.yml & my.yml

**Comment by ainulfarizah** [comment]
ready

File: app/views/layouts/_header.html.erb [hunk 5]
```diff
@@ -3,9 +3,29 @@
...
+            <%= link_to "#" class: "dropdown-toggle" "data-toggle": "dropdown" do %>
+              <%= t "account" %>
+              <b class="caret"></b>
+            <% end %>
+          <ul class="dropdown-menu">
+            <li><%= link_to "Profile", current_user %></li>
+            <li><%= link_to "Settings", '#' %></li>
+            <li class="divider"></li>
+            <li> <%= link_to t(:log_out), logout_path, method: :delete %></li>
+          </ul>
```
Comment (ghost) [inline app/views/layouts/_header.html.erb]: indent please

File: app/views/sessions/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+<h1>Log in</h1>
```
Comment (ghost) [inline app/views/sessions/new.html.erb]: i18n please

**Review by ghost — COMMENTED** [review]
cmt

**Comment by ainulfarizah** [comment]
ready

File: app/helpers/sessions_helper.rb [hunk 7]
```diff
@@ -10,12 +10,19 @@ def current_user
...
+      @user = User.find_by id: user_id
       if user&.authenticated?(cookies[:remember_token])
-        log_in user
-        @current_user = user
+        log_in @user
+        @current_user = @user
       end
     end
+
+  session[:user_id] = user.id
```
Comment (huongnt-2545) [inline app/helpers/sessions_helper.rb]: remove this line

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -14,6 +14,7 @@ def new
   def create
     @user = User.new(user_params)
     if @user.save  # Handle a successful save.
+      login_in @user
```
Comment (huongnt-2545) [inline app/controllers/users_controller.rb]:
```suggestion
      log_in @user
```

File: config/routes.rb [hunk 9]
```diff
@@ -1,8 +1,15 @@
 Rails.application.routes.draw do
   scope "(:locale)", locale: /#{I18n.available_locales.join("|")}/ do
-  root 'static_pages#home'
```
Comment (huongnt-2545) [inline config/routes.rb:3]: keep line 3, 4

File: config/routes.rb [hunk 10]
```diff
@@ -1,8 +1,15 @@
...
-  get 'static_pages/help'
+
+  get 'sessions/create'
+  get 'sessions/destroy'
+  root :to => 'static_pages#home', as: "home"
```
Comment (huongnt-2545) [inline config/routes.rb]: remove line 6, 7

**Comment by ainulfarizah** [comment]
ready

File: app/helpers/sessions_helper.rb [hunk 11]
```diff
@@ -10,14 +10,19 @@ def current_user
     if (user_id = session[:user_id])
       @current_user ||= User.find_by id: user_id
     elsif (user_id = cookies.signed[:user_id])
-      user = User.find_by id: user_id
+      @user = User.find_by id: user_id
       if user&.authenticated?(cookies[:remember_token])
```
Comment (huongnt-2545) [inline app/helpers/sessions_helper.rb]:
```suggestion
      if @user&.authenticated?(cookies[:remember_token])
```

File: app/helpers/users_helper.rb [hunk 12]
```diff
@@ -1,8 +1,7 @@
 module UsersHelper
 
-  # Returns the Gravatar for the given user.
   def gravatar_for user
-    gravatar_id = Digest::MD5::hexdigest user.email.downcase
```
Comment (huongnt-2545) [inline app/helpers/users_helper.rb:5]: keep this

File: app/models/user.rb [hunk 13]
```diff
@@ -5,6 +5,16 @@ class User < ApplicationRecord
...
+  def User.digest string
+    cost = if ActiveModel::SecurePassword.min_cost
+             BCrypt::Engine::MIN_COST
+           else
+             BCrypt::Engine.cost
+           end
+    BCrypt::Password.create string, cost: cost
+  end
+
   has_secure_password
```
Comment (huongnt-2545) [inline app/models/user.rb]: move this line place to below line 6

**Comment by ainulfarizah** [comment]
ready

**Review by ghost — COMMENTED** [review]
merged

File: config/routes.rb [hunk 14]
```diff
@@ -1,8 +1,15 @@
...
   get 'static_pages/help'
   get "/signup",  to: "users#new"
   post "/signup", to:"users#create"
-  resources :users, only: %i(new create show)
+  resources :users, only: %i(show new create)
+
+  get "/login", to: "sessions#new"
+  post "/login", to: "sessions#create"
+  delete "/logout", to: "sessions#destroy"
 end
```
Comment (huongnt-2545) [inline config/routes.rb:13]: Add the `end` keyword at the end of file

File: config/routes.rb [hunk 15]
```diff
@@ -1,8 +1,15 @@
 Rails.application.routes.draw do
   scope "(:locale)", locale: /#{I18n.available_locales.join("|")}/ do
+
+  get 'sessions/create'
```
Comment (huongnt-2545) [inline config/routes.rb]: remove line 4, 5

File: config/routes.rb (hunk 14 above)
Comment (ainulfarizah) [reply config/routes.rb:13]: Screenshot from 2021-09-08 11-19-54

i already have

**Pull request merged**